scripts/migrateAttendance.js

```diff
@@ -0,0 +1,179 @@
+// to a new attendance collection
+
+// Standardize date to midnight
+function standardDate(d){
+    var date = new Date(d);
+    date.setHours(0,0,0,0);
+    return date;
+}
+
+var smallgroups = db.smallgroups.find({});
+// Remove the "semester" string in all the small groups and replace
+// with the correct "classyear" id
+while(smallgroups.hasNext()) {
+    smallgroups.forEach(function(group){
+        if (group.semester){
+            var classyear = db.classyears.findOne({semester:group.semester})
+            db.smallgroups.update(
+                { "_id": group._id },
+                {
+                    $set: { "classYear": classyear._id },
+                    $unset: { semester: 1 }
+                }
+            );
+        }
+    })
+
+}
+
+var users = db.users.find({});
+var classyear = db.classyears.findOne({current:true});
+
+while(users.hasNext()){
+    //Get all the users
+    users.forEach(function(user){
+        var attendance = user.attendance;
+        var unverifiedAttendance = user.unverifiedAttendance;
+        var smallgroup = false;
+        //Get the user's small group
+        if (user.smallgroup){
+            smallgroup = db.smallgroups.findOne({"_id":user.smallgroup});
+        }
+        var allfound = true;
+        //Loop through all the attendance entries
+        if (attendance){
+            for (var a = 0; a < attendance.length ; a++){
+                var date = attendance[a];
+                var found  = false;
+                //Find the correct attendance code in the classyear
+                for (var b = 0; b < classyear.dayCodes.length ; b++){
+                    if (standardDate(classyear.dayCodes[b].date).getTime() == standardDate(date).getTime()){
+                        var daycode = classyear.dayCodes[b];
+                        db.attendances.insert(
+                            {
+                                classYear: classyear._id,
+                                user: user._id,
+
+                                date: standardDate(date),
+                                datetime: date,
+
+                                bonusDay: daycode.bonusDay||false,
+                                smallgroup: false,
+
+                                verified: true,
+                                code: daycode.code,
+                            }
+                        );
+                        found = true;
+                        break;
+                    }
+                }
+                //If not found, find the correct attendance code in the smallgroup
+                //Check to make sure there is a smallgroup for the user with daycodes
+                if (!found && user.smallgroup && smallgroup && smallgroup.dayCodes){
+                    for (var c = 0; c < smallgroup.dayCodes.length ; c++){
+                        if (standardDate(smallgroup.dayCodes[c].date).getTime() == standardDate(date).getTime()){
```

Comment:
this same long line is used in multiple places. why not make it a function that returns a boolean. would be easier to understand and update later.